src/gtc/passes/oir_pipeline.py

```diff
@@ -50,18 +51,32 @@ def hash_step(step: Callable) -> int:
     return hash(step)
 
 
-class OirPipeline:
+class OirPipeline(ABC):
+    @abstractmethod
+    def __hash__(self) -> int:
+        pass
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        pass
+
+    @abstractmethod
+    def run(self, oir: oir.Stencil) -> oir.Stencil:
+        pass
+
```

Comment:
I suggest to make `OirPipeline` a `typing.Protocol` which is also an `ABC`, so it is possible to define `abstractmethod`s.

I would also suggest to add a `raise NotImplementedError("Missing implementation of ....")` to the body of the abstract methods, since they can be call by mistake from the subclasses via `super()`